### DISCUSSION TITLE
brew is asking for git credentials when searching for a package

### DISCUSSION THREAD

**Discussion by thomasdziedzic-calmwave** [Homebrew/discussions#4128] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 3.6.20
ORIGIN: https://github.com/Homebrew/brew
HEAD: 344d32bf7f40183359592ef81ca6486e0587981b
Last commit: 4 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 29d36a32a689c2ccfbdc5eddbd578cb7d3337e05
Core tap last commit: 42 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: code --wait
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.39.1 => /opt/homebrew/bin/git
Curl: 7.85.0 => /usr/bin/curl
macOS: 13.1-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

After running `brew search -dv gmp` I get a prompt, which says:

![Screenshot 2023-01-23 at 8 01 36 AM](https://user-images.githubusercontent.com/112598295/214058559-a6cf910f-4efc-4b3d-93d6-0dfde9054d11.png)

> git-credential-osxkeychain wants to use your confidential information stored in "github.com" in your keychain

After I deny the prompt twice, I get the following output:

```
/opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/3.6.20\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 13.1\)\ curl/7.85.0 --header Accept-Language:\ en --retry 3 --location https://api.github.com/search/code\?q=user\%3AHomebrew\+path\%3AFormula\+path\%3ACasks\+path\%3A.\+filename\%3Agmp\+extension\%3Arb\&per_page=100 --header Accept:\ application/vnd.github\+json --write-out '
'\%\{http_code\} --header X-GitHub-Api-Version:2022-11-28 --dump-header /private/tmp/github_api_headers20230123-13014-l9lfdu
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/gmt.rb
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/gvp.rb
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/gpp.rb
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/amp.rb
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/emp.rb
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/xmp.rb
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/gmp.rb
```

Why is it asking for my git credentials?

**Comment by SMillerDev** [comment, marked as answer]
Homebrew software repositories are git reposetories that are hosted on GitHub. To search it uses a combination of your local files and the GitHub API.

**Reply by thomasdziedzic-calmwave** [reply]
But I am using publicly available taps:

```
% brew tap
homebrew/core
homebrew/services
```

Since brew isn't aware of any private taps, I wouldn't expect it to ask me for my git credentials.

**Reply by SMillerDev** [reply]
It's using the GitHub API with your default credentials to ensure you don't hit a rate limit when you're trying to do something.

**Reply by thomasdziedzic-calmwave** [reply]
But the search is only searching my local disk space afaict because when I deny the prompt, it still returns the correct results. Or is it attempting to search remotely and when it's denied access, it will fallback to a local search?

**Reply by SMillerDev** [reply]
That's the general idea, yes.

**Reply by thomasdziedzic-calmwave** [reply]
Got it, thanks for the explanation!

Is this a relatively new change?

Would it make sense to check if there is a token available in the environment like https://github.com/Homebrew/brew/pull/14349/files instead of going into the keychain for the github token and if there is none specified to just search the local clones of taps?

**Reply by SMillerDev** [reply]
As far as I know it's been there for about 5 years. If you have suggestions for improvements, please make a pull request.
